File: datamart/src/main/scala/datamart.scala
import org.apache.spark.sql.{DataFrame, SparkSession}
import preprocess.{Preprocess}
import db.{Database}
import com.typesafe.scalalogging.Logger


class DataMart(HOST: String) {
  private val USER = "root"
  private val PASSWORD = "password"
  private val APP_NAME = "KMeans"
  private val DEPLOY_MODE = "local"
  private val DRIVER_MEMORY = "2g"
  private val EXECUTOR_MEMORY = "2g"
  private val EXECUTOR_CORES = 1
  private val DRIVER_CORES = 1
  private val POSTGRESQL_CONNECTOR_JAR = "../postgresql-42.7.3.jar"
  val session = SparkSession.builder
    .appName(APP_NAME)
    .master(DEPLOY_MODE)
    .config("spark.driver.host", HOST)
    .config("spark.driver.bindAddress", HOST)
    .config("spark.driver.cores", DRIVER_CORES)
    .config("spark.executor.cores", EXECUTOR_CORES)
    .config("spark.driver.memory", DRIVER_MEMORY)
    .config("spark.executor.memory", EXECUTOR_MEMORY)
    .config("spark.jars", POSTGRESQL_CONNECTOR_JAR)
    .config("spark.driver.extraClassPath", POSTGRESQL_CONNECTOR_JAR)
    .getOrCreate()
  private val db = new Database(session, HOST)
  private val logger = Logger("Logger")


  def read_dataset(tablename: String): DataFrame = {
    val data = db.read_table(tablename)
    logger.info("The table was successfully read")

    val transforms: Seq[DataFrame => DataFrame] = Seq(
      Preprocess.fillNa,
      Preprocess.assemble_vector,
      Preprocess.stdize_assembled_dataset
    )

    val transformed = transforms.foldLeft(data) { (df, f) => f(df) }
    logger.info("All transforms were applied to the dataset")
    transformed
  }

  def write_predictions(df: DataFrame): Unit = {
    db.insert_df(df, "Predictions")
    logger.info("All predictions were inserted in the Predictions table")
  }
}

File: src/datamart.py
from pyspark.sql import DataFrame, SparkSession, SQLContext
from pyspark.sql.types import StructField, IntegerType, StructType
from pyspark.ml.linalg import VectorUDT
from logger import Logger
import os

SHOW_LOG = True

class Datamart:
    def __init__(self, spark: SparkSession, args):
        #HOST,PORT,DB,USER,PASS,TABLENAME, ((MODE))
        self.spark = spark
        self.host = args[0]
        self.port = args[1]
        self.db = args[2]
        self.user = args[3]
        self.passw = args[4]
        self.table = args[5]

        logger = Logger(SHOW_LOG)
        self.log = logger.get_logger(__name__)

        self.schema = StructType([
                StructField("id", IntegerType(), True),
                StructField("stdized_features", VectorUDT(), True),
        ])

        os.system(f'spark-shell --jars postgresql-42.7.3.jar -i datamart\datamart.scala --conf spark.driver.args="{self.host},{self.port},{self.db},{self.user},{self.passw},{self.table},read"')
        self.log.info("Datamart initialized with preprocessed json data saved to data/datamart")


    def read_dataset(self) -> DataFrame:
        df = self.spark.read.json('data/datamart/*.json', self.schema)
        return df

    def write_predictions(self, df: DataFrame):
        df.write.mode("overwrite").json('data/predictions')
        os.system(f'spark-shell --jars postgresql-42.7.3.jar -i datamart\datamart.scala --conf spark.driver.args="{self.host},{self.port},{self.db},{self.user},{self.passw},Predictions,write"')